{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# xylE dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-12-25T20:48:32.721016Z",
     "start_time": "2021-12-25T20:48:30.929057Z"
    }
   },
   "outputs": [],
   "source": [
    "# Standard imports\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "# Special imports\n",
    "import mavenn\n",
    "import pyfasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Summary \n",
    "\n",
    "Sort-seq MPRA data from Belliveau et. al., 2018. The authors used fluoresence-activated cell sorting of *E. coli* cells, followed by deep sequencing, to assay gene expression levels from variant  *xylE* promoters. Specifically, the authors created three plasmid libraries containing variant *xylE* promoters driving GFP expression. For each library, *xylE* promoters were mutagenized within a window of 52 bp (library 1), 53 bp (library 2), or 60 bp (libary 3) in length, after which cells were sorted into 4 bins using FACS. The three mutagenized windows partially overlap and cover a total of 150 bp spanning positions [-127:+23] relative to the primary *xylE* transcription start site.\n",
    "\n",
    "Below we format data files provided by Belliveau et al. into a format suitable for analysis in MAVE-NN. In the final formatted dataset, column `'x'` lists variant 150 bp DNA sequences (each containing the variable region embedded as appropriate within the larger wild-type sequence), columns `'ct_0'` to `'ct_11'` list the number of reads obtained for each sequence in each of the 12 FACS bins (4 bins per library), and the `'set'` column indicates whether each sequence is to assigned to the training set, the validation set, or the test set.\n",
    "\n",
    "**Name**: ``'xylE'``\n",
    "\n",
    "**Reference**: Belliveau NM, Barnes SL, Ireland WT, Jones DL, Sweredoski MJ, Moradian A, Hess S, Kinney JB, Phillips R. Systematic approach for dissecting the molecular mechanisms of transcriptional regulation in bacteria. [Proc Natl Acad Sci USA 115:E4796–E4805 (2018).](https://doi.org/10.1073/pnas.1722055115)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-12-25T20:48:35.797042Z",
     "start_time": "2021-12-25T20:48:32.722151Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>set</th>\n",
       "      <th>ct_0</th>\n",
       "      <th>ct_1</th>\n",
       "      <th>ct_2</th>\n",
       "      <th>ct_3</th>\n",
       "      <th>ct_4</th>\n",
       "      <th>ct_5</th>\n",
       "      <th>ct_6</th>\n",
       "      <th>ct_7</th>\n",
       "      <th>ct_8</th>\n",
       "      <th>ct_9</th>\n",
       "      <th>ct_10</th>\n",
       "      <th>ct_11</th>\n",
       "      <th>x</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>training</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>AAACAATAGCATTGTTCTTATCAATTTTGGATAAGTATTATAATTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>test</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>AAACAATAGTATCGTTTTTAGCTCATTTGGATAATTATTACAATTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>test</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>AAACAATAGTATTGCTTTTATCAATTTAGGATAATTATCACAATTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>training</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>AAACAATAGTATTGCTTTTATCAATTTAGGATAATTATCACGATTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>training</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>AAACAATAGTATTGCTTTTATCATTTTTGGAAAATTAACACGATTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1847223</th>\n",
       "      <td>training</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>TTTCAATAGTCTTGTTTTTATCAATTTTGGATAATTATCACAATTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1847224</th>\n",
       "      <td>validation</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>TTTCACTAGAATTGTTTTTATCAATTGTGGATAATTTTCACAATTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1847225</th>\n",
       "      <td>training</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>TTTCACTAGAATTGTTTTTATCAATTGTGGATAATTTTCACAATTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1847226</th>\n",
       "      <td>validation</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>TTTCAGTAGTATGGTTTTTATCAATTTTGGATAATTATCACATTTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1847227</th>\n",
       "      <td>training</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>TTTCAGTAGTATTGTTTTTATCAATTTTGGATAATTATCACATTTA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1847228 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                set  ct_0  ct_1  ct_2  ct_3  ct_4  ct_5  ct_6  ct_7  ct_8  \\\n",
       "0          training     0     0     0     0     0     0     0     0     0   \n",
       "1              test     0     0     0     0     0     0     0     0     0   \n",
       "2              test     0     0     0     0     0     0     0     0     0   \n",
       "3          training     0     0     0     0     0     0     0     0     0   \n",
       "4          training     0     0     0     0     0     0     0     0     0   \n",
       "...             ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   \n",
       "1847223    training     0     0     0     0     0     0     0     0     1   \n",
       "1847224  validation     0     0     0     0     0     0     0     0     0   \n",
       "1847225    training     0     0     0     0     0     0     0     0     0   \n",
       "1847226  validation     0     0     0     0     0     0     0     0     0   \n",
       "1847227    training     0     0     0     0     0     0     0     0     0   \n",
       "\n",
       "         ct_9  ct_10  ct_11                                                  x  \n",
       "0           0      0      3  AAACAATAGCATTGTTCTTATCAATTTTGGATAAGTATTATAATTA...  \n",
       "1           0      1      0  AAACAATAGTATCGTTTTTAGCTCATTTGGATAATTATTACAATTA...  \n",
       "2           0      0      1  AAACAATAGTATTGCTTTTATCAATTTAGGATAATTATCACAATTA...  \n",
       "3           0      0      5  AAACAATAGTATTGCTTTTATCAATTTAGGATAATTATCACGATTA...  \n",
       "4           0      0      2  AAACAATAGTATTGCTTTTATCATTTTTGGAAAATTAACACGATTA...  \n",
       "...       ...    ...    ...                                                ...  \n",
       "1847223     0      0      0  TTTCAATAGTCTTGTTTTTATCAATTTTGGATAATTATCACAATTT...  \n",
       "1847224     0      1      0  TTTCACTAGAATTGTTTTTATCAATTGTGGATAATTTTCACAATTA...  \n",
       "1847225     0      1      0  TTTCACTAGAATTGTTTTTATCAATTGTGGATAATTTTCACAATTA...  \n",
       "1847226     0      3      0  TTTCAGTAGTATGGTTTTTATCAATTTTGGATAATTATCACATTTA...  \n",
       "1847227     0      1      0  TTTCAGTAGTATTGTTTTTATCAATTTTGGATAATTATCACATTTA...  \n",
       "\n",
       "[1847228 rows x 14 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mavenn.load_example_dataset('xylE')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "# Preprocessing\n",
    "\n",
    "*xylE* is transcribed in the reverse direction starting from a  position 4,242,291 in the *E. coli* genome. Belliveau et al. studied a 150 bp region of the *xylE* promoter, spanning genomic coordinates [4,242,269:4,242,418]. We designate the reverse-strand-sequence of this genomic region as the wild-type sequence. Within this 150 bp region, library 1 spans positions [98:150], library 2 spans positions [52:105], and library 3 spans positions [1:60]. A schematic is shown below.\n",
    "\n",
    "![Image of 150 bp region.](xylE_promoter.png)\n",
    "\n",
    "We begin by extracting the 150 bp wild-type sequence from the *E. coli* genome. Note that this step requires the `pyfasta` Python package, which is not a dependency of MAVE-NN."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-12-25T20:48:35.841541Z",
     "start_time": "2021-12-25T20:48:35.798112Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wild-type sequence (length 150 bp):\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'CGGCAATAGTATTGTTTTTATCAATTTTGGATAATTATCACAATTAAGATCACAGAAAAGACATTACGTAAACGCATTGTAAAAAATGATAATTGCCTTAACTGCCTGACAATTCCAACATCAATGCACTGATAAAAGATCAGAATGGTC'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Extract WT sequence from E. coli genome\n",
    "# Genome FASTA file downloaded from https://www.ncbi.nlm.nih.gov/assembly/GCF_000005845.2/ on 12/25/2021\n",
    "genome_file = '../../mavenn/examples/datasets/raw/GCF_000005845.2_ASM584v2_genomic.fna'\n",
    "genome = pyfasta.Fasta(genome_file)\n",
    "wt_seq = genome.sequence({'chr':'NC_000913.3 Escherichia coli str. K-12 substr. MG1655, complete genome',\n",
    "                          'start': 4242269,\n",
    "                          'stop': 4242418,\n",
    "                          'strand': '-'})\n",
    "print(f'Wild-type sequence (length {len(wt_seq)} bp):')\n",
    "wt_seq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next we load and concatenate read counts for each of the three libraries assayed by Belliveau et al.. In the resulting dataframe, `raw_data_df`, each row corresponds to a different Illumina read. Column `'x`' lists only the sequence of the mutagenized promoter region, column `'bin'` lists the FACS bin (0, 1, 2 , or 3) that read was found in, and `'lib'` indicates the library that read was derived from. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-12-25T20:48:40.461744Z",
     "start_time": "2021-12-25T20:48:35.842358Z"
    },
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading lib1 data from ../../mavenn/examples/datasets/raw/xylE_lib1_raw.csv.gz...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/jkinney/miniforge3_arm64/lib/python3.9/site-packages/numpy/lib/arraysetops.py:580: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison\n",
      "  mask |= (ar1 == a)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading lib2 data from ../../mavenn/examples/datasets/raw/xylE_lib2_raw.csv.gz...\n",
      "Loading lib3 data from ../../mavenn/examples/datasets/raw/xylE_lib3_raw.csv.gz...\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>x</th>\n",
       "      <th>bin</th>\n",
       "      <th>lib</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TTAACTGTCTGACGATTCAAACATCAATACACTAATAAAAGATCAG...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TTAACTGCATGACAATTCCAACCTCAATGCATTGATAAAAGATCAG...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TTAACTGCCTGACAATTCCAAGATCAATGCAGTGATAAAGGATCAG...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TTTACTGCGTGTCAATTCGGGCAGCAGTACACTTATAAGAGATCAG...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TAAACTACCTGACAACTCCAACTTTAACGCACTGATTACAGTTCAG...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6859154</th>\n",
       "      <td>CGGCAATAGTATTGTTTTTATCAATATTGGATAATTAACACAATTA...</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6859155</th>\n",
       "      <td>CGGCTATAGTATTGTTTTTATCAATTTTGGATAATTATCACAGTTA...</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6859156</th>\n",
       "      <td>CGGCAATAGTATGGTTTTTATCAATTTTGGACAATTATCAAAATTC...</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6859157</th>\n",
       "      <td>CAGCAATAGTATTGTTTTTATCAATTTTTGATAATTATCCCAATTA...</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6859158</th>\n",
       "      <td>GGGCGATAGTATAGTTTTTATCAATTTTGGATAATTATTTCAATTA...</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6859159 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                         x bin lib\n",
       "0        TTAACTGTCTGACGATTCAAACATCAATACACTAATAAAAGATCAG...   0   1\n",
       "1        TTAACTGCATGACAATTCCAACCTCAATGCATTGATAAAAGATCAG...   0   1\n",
       "2        TTAACTGCCTGACAATTCCAAGATCAATGCAGTGATAAAGGATCAG...   0   1\n",
       "3        TTTACTGCGTGTCAATTCGGGCAGCAGTACACTTATAAGAGATCAG...   0   1\n",
       "4        TAAACTACCTGACAACTCCAACTTTAACGCACTGATTACAGTTCAG...   0   1\n",
       "...                                                    ...  ..  ..\n",
       "6859154  CGGCAATAGTATTGTTTTTATCAATATTGGATAATTAACACAATTA...   3   3\n",
       "6859155  CGGCTATAGTATTGTTTTTATCAATTTTGGATAATTATCACAGTTA...   3   3\n",
       "6859156  CGGCAATAGTATGGTTTTTATCAATTTTGGACAATTATCAAAATTC...   3   3\n",
       "6859157  CAGCAATAGTATTGTTTTTATCAATTTTTGATAATTATCCCAATTA...   3   3\n",
       "6859158  GGGCGATAGTATAGTTTTTATCAATTTTGGATAATTATTTCAATTA...   3   3\n",
       "\n",
       "[6859159 rows x 3 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Download libraries and embed variant sequences within\n",
    "data_file_template = '../../mavenn/examples/datasets/raw/xylE_lib{}_raw.csv.gz'\n",
    "raw_data_df = pd.DataFrame(columns=['x','bin','lib'])\n",
    "for lib_num in [1,2,3]:\n",
    "\n",
    "    # Read in dataset\n",
    "    data_file = data_file_template.format(lib_num)\n",
    "    print(f'Loading lib{lib_num} data from {data_file}...')\n",
    "    lib_data_df= pd.read_csv(data_file,\n",
    "                             compression='gzip',\n",
    "                             index_col=[0])\n",
    "    lib_data_df['lib'] = lib_num\n",
    "    raw_data_df = pd.concat([raw_data_df, lib_data_df],\n",
    "                            ignore_index=True)\n",
    "\n",
    "raw_data_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next we pad each variant sequence in `raw_data_df` with wild-type sequence as appropriate. We also renumber the bins so that bins 0-3 correspond to library 1, bins 4-7 correspond to library 2, and bins 8-11 correspond to library 3. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-12-25T20:48:46.048899Z",
     "start_time": "2021-12-25T20:48:40.462628Z"
    },
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Verified: all sequences are now of length 150.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>x</th>\n",
       "      <th>bin</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CGGCAATAGTATTGTTTTTATCAATTTTGGATAATTATCACAATTA...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CGGCAATAGTATTGTTTTTATCAATTTTGGATAATTATCACAATTA...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CGGCAATAGTATTGTTTTTATCAATTTTGGATAATTATCACAATTA...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CGGCAATAGTATTGTTTTTATCAATTTTGGATAATTATCACAATTA...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CGGCAATAGTATTGTTTTTATCAATTTTGGATAATTATCACAATTA...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6859154</th>\n",
       "      <td>CGGCAATAGTATTGTTTTTATCAATATTGGATAATTAACACAATTA...</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6859155</th>\n",
       "      <td>CGGCTATAGTATTGTTTTTATCAATTTTGGATAATTATCACAGTTA...</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6859156</th>\n",
       "      <td>CGGCAATAGTATGGTTTTTATCAATTTTGGACAATTATCAAAATTC...</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6859157</th>\n",
       "      <td>CAGCAATAGTATTGTTTTTATCAATTTTTGATAATTATCCCAATTA...</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6859158</th>\n",
       "      <td>GGGCGATAGTATAGTTTTTATCAATTTTGGATAATTATTTCAATTA...</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6859159 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                         x bin\n",
       "0        CGGCAATAGTATTGTTTTTATCAATTTTGGATAATTATCACAATTA...   0\n",
       "1        CGGCAATAGTATTGTTTTTATCAATTTTGGATAATTATCACAATTA...   0\n",
       "2        CGGCAATAGTATTGTTTTTATCAATTTTGGATAATTATCACAATTA...   0\n",
       "3        CGGCAATAGTATTGTTTTTATCAATTTTGGATAATTATCACAATTA...   0\n",
       "4        CGGCAATAGTATTGTTTTTATCAATTTTGGATAATTATCACAATTA...   0\n",
       "...                                                    ...  ..\n",
       "6859154  CGGCAATAGTATTGTTTTTATCAATATTGGATAATTAACACAATTA...  11\n",
       "6859155  CGGCTATAGTATTGTTTTTATCAATTTTGGATAATTATCACAGTTA...  11\n",
       "6859156  CGGCAATAGTATGGTTTTTATCAATTTTGGACAATTATCAAAATTC...  11\n",
       "6859157  CAGCAATAGTATTGTTTTTATCAATTTTTGATAATTATCCCAATTA...  11\n",
       "6859158  GGGCGATAGTATAGTTTTTATCAATTTTGGATAATTATTTCAATTA...  11\n",
       "\n",
       "[6859159 rows x 2 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# List beginning and end of each variant region, in Python coordinates\n",
    "lib_coords_dict = {\n",
    "    1: [97, 150],\n",
    "    2: [51, 105],\n",
    "    3: [0, 60]\n",
    "}\n",
    "\n",
    "tidy_data_df = raw_data_df[['x','bin']].copy()\n",
    "for lib_num, (start, stop) in lib_coords_dict.items():\n",
    "\n",
    "    # Get indices for sequences in specific library\n",
    "    ix = (raw_data_df['lib']==lib_num)\n",
    "\n",
    "    # Those variant sequences sequence into larger WT sequence\n",
    "    tidy_data_df.loc[ix, 'x'] = [\n",
    "        wt_seq[:start] + x + wt_seq[stop:]\n",
    "        for x in tidy_data_df.loc[ix,'x']\n",
    "    ]\n",
    "\n",
    "# Verify that all sequences are the correct length\n",
    "assert np.all([len(x)==len(wt_seq) for x in tidy_data_df['x']])\n",
    "print(f'Verified: all sequences are now of length {len(wt_seq)}.')\n",
    "\n",
    "# Alter bin numbers to accommodate 12 bins\n",
    "tidy_data_df['bin'] = raw_data_df['bin'] + 4*(raw_data_df['lib']-1)\n",
    "\n",
    "tidy_data_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally, we convert these data from the \"tidy\" format in `tidy_data_df` to a \"wide\" format in which each row corresponds to a different sequence and the read counts for each bin are listed in a separate column. This wide format is then leveraged to assign each unique variant promoter to the `training`, `validatiaon`, or `test` set. The final dataset is then saved to file. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-12-25T20:48:58.694777Z",
     "start_time": "2021-12-25T20:48:46.049661Z"
    },
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>set</th>\n",
       "      <th>ct_0</th>\n",
       "      <th>ct_1</th>\n",
       "      <th>ct_2</th>\n",
       "      <th>ct_3</th>\n",
       "      <th>ct_4</th>\n",
       "      <th>ct_5</th>\n",
       "      <th>ct_6</th>\n",
       "      <th>ct_7</th>\n",
       "      <th>ct_8</th>\n",
       "      <th>ct_9</th>\n",
       "      <th>ct_10</th>\n",
       "      <th>ct_11</th>\n",
       "      <th>x</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>training</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>AAACAATAGCATTGTTCTTATCAATTTTGGATAAGTATTATAATTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>test</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>AAACAATAGTATCGTTTTTAGCTCATTTGGATAATTATTACAATTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>test</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>AAACAATAGTATTGCTTTTATCAATTTAGGATAATTATCACAATTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>training</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>AAACAATAGTATTGCTTTTATCAATTTAGGATAATTATCACGATTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>training</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>AAACAATAGTATTGCTTTTATCATTTTTGGAAAATTAACACGATTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1847223</th>\n",
       "      <td>training</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>TTTCAATAGTCTTGTTTTTATCAATTTTGGATAATTATCACAATTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1847224</th>\n",
       "      <td>validation</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>TTTCACTAGAATTGTTTTTATCAATTGTGGATAATTTTCACAATTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1847225</th>\n",
       "      <td>training</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>TTTCACTAGAATTGTTTTTATCAATTGTGGATAATTTTCACAATTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1847226</th>\n",
       "      <td>validation</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>TTTCAGTAGTATGGTTTTTATCAATTTTGGATAATTATCACATTTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1847227</th>\n",
       "      <td>training</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>TTTCAGTAGTATTGTTTTTATCAATTTTGGATAATTATCACATTTA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1847228 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                set  ct_0  ct_1  ct_2  ct_3  ct_4  ct_5  ct_6  ct_7  ct_8  \\\n",
       "0          training     0     0     0     0     0     0     0     0     0   \n",
       "1              test     0     0     0     0     0     0     0     0     0   \n",
       "2              test     0     0     0     0     0     0     0     0     0   \n",
       "3          training     0     0     0     0     0     0     0     0     0   \n",
       "4          training     0     0     0     0     0     0     0     0     0   \n",
       "...             ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   \n",
       "1847223    training     0     0     0     0     0     0     0     0     1   \n",
       "1847224  validation     0     0     0     0     0     0     0     0     0   \n",
       "1847225    training     0     0     0     0     0     0     0     0     0   \n",
       "1847226  validation     0     0     0     0     0     0     0     0     0   \n",
       "1847227    training     0     0     0     0     0     0     0     0     0   \n",
       "\n",
       "         ct_9  ct_10  ct_11                                                  x  \n",
       "0           0      0      3  AAACAATAGCATTGTTCTTATCAATTTTGGATAAGTATTATAATTA...  \n",
       "1           0      1      0  AAACAATAGTATCGTTTTTAGCTCATTTGGATAATTATTACAATTA...  \n",
       "2           0      0      1  AAACAATAGTATTGCTTTTATCAATTTAGGATAATTATCACAATTA...  \n",
       "3           0      0      5  AAACAATAGTATTGCTTTTATCAATTTAGGATAATTATCACGATTA...  \n",
       "4           0      0      2  AAACAATAGTATTGCTTTTATCATTTTTGGAAAATTAACACGATTA...  \n",
       "...       ...    ...    ...                                                ...  \n",
       "1847223     0      0      0  TTTCAATAGTCTTGTTTTTATCAATTTTGGATAATTATCACAATTT...  \n",
       "1847224     0      1      0  TTTCACTAGAATTGTTTTTATCAATTGTGGATAATTTTCACAATTA...  \n",
       "1847225     0      1      0  TTTCACTAGAATTGTTTTTATCAATTGTGGATAATTTTCACAATTA...  \n",
       "1847226     0      3      0  TTTCAGTAGTATGGTTTTTATCAATTTTGGATAATTATCACATTTA...  \n",
       "1847227     0      1      0  TTTCAGTAGTATTGTTTTTATCAATTTTGGATAATTATCACATTTA...  \n",
       "\n",
       "[1847228 rows x 14 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Convert data in \"tidy\" format to \"wide\" format\n",
    "ct_my, x_m = mavenn.src.utils.vec_data_to_mat_data(\n",
    "    y_n=tidy_data_df['bin'],\n",
    "    x_n=tidy_data_df['x']\n",
    ")\n",
    "M, Y = ct_my.shape\n",
    "ct_cols = [f'ct_{y}' for y in range(Y)]\n",
    "wide_data_df = pd.DataFrame()\n",
    "wide_data_df[ct_cols] = ct_my\n",
    "wide_data_df['x'] = x_m\n",
    "\n",
    "# Assign sequences to training, validation, and test sets\n",
    "np.random.seed(0)\n",
    "sets = np.random.choice(a=['training','test','validation'],\n",
    "                        p=[.6,.2,.2], \n",
    "                        size=len(wide_data_df))\n",
    "wide_data_df.insert(loc=0, column='set', value=sets)\n",
    "\n",
    "# # Save dataset (uncomment to execute)\n",
    "# wide_data_df.to_csv('xylE_data.csv.gz',index=False, compression='gzip')\n",
    "\n",
    "wide_data_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-12-25T20:46:20.065813Z",
     "start_time": "2021-12-25T20:46:20.062217Z"
    }
   },
   "source": [
    "This final dataframe, `wide_data_df`, has the same format as the `'xylE'` dataset that comes with MAVE-NN. "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.2"
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
